xmake/rules/qt/qmltyperegistrar/xmake.lua

```diff
@@ -0,0 +1,125 @@
+--!A cross-platform build utility based on Lua
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      arthapz
+-- @file        xmake.lua
+--
+
+rule("qt.qmltyperegistrar")
+    add_deps("qt.env")
+    set_extensions(".h", ".hpp")
+
+    on_load(function(target)
+        -- get qt
+        local qt = assert(target:data("qt"), "qt not found!")
+
+        -- get qmltyperegistrar
+        local qmltyperegistrar = path.join(qt.bindir, is_host("windows") and "qmltyperegistrar.exe" or "qmltyperegistrar")
+        if not os.isexec(qmltyperegistrar) and qt.libexecdir then
+            qmltyperegistrar = path.join(qt.libexecdir, is_host("windows") and "qmltyperegistrar.exe" or "qmltyperegistrar")
+        end
+        if not os.isexec(qmltyperegistrar) and qt.libexecdir_host then
+            qmltyperegistrar = path.join(qt.libexecdir_host, is_host("windows") and "qmltyperegistrar.exe" or "qmltyperegistrar")
+        end
+        assert(qmltyperegistrar and os.isexec(qmltyperegistrar), "qmltyperegistrar not found!")
+
+        -- set targetdir
+        local importname = target:values("qt.qmlplugin.import_name")
+        assert(importname, "QML plugin import name not set")
+
+        local targetdir = path.join(target:targetdir(), "imports")
+        for _, dir in pairs(importname:split(".", { plain = true })) do
+            targetdir = path.join(targetdir, dir)
+        end
+        os.mkdir(targetdir)
+        target:set("targetdir", targetdir)
+
+        -- add qmldir
+        local qmldir = target:values("qt.qmlplugin.qmldir")
+        if qmldir then
+            target:add("installfiles", qmldir)
```

Comment:
installfiles is directory? I don't remember supporting it.

Reply:
qmldir is not a directory, this is a file

Reply:
(and it has to be named qmldir, not my name)

Reply:
why?

Reply:
qt stuffs

Reply:
I think it will be better of we use `qt.qmlplugin.qmldirfile`

Reply:
Okay